mantidproject/mantid
# Crystalfield bugfixes

This is a collection of bugfixes for the CrystalField calculations routines since the back-end was rewritten last autumn. Due to the changes in the back-end some workflows (which don't have system tests) became broken, and this was not picked up in the unit tests.

The major changes are in the handling of peak widths (setting a resolution model disables peak with fitting and invalidates the `.FWHM` property; setting `.FWHM` invalidates a resolution model; previously the behaviour when both are set was ambiguous and in some cases wrong (e.g. fixed FWHM at starting value given in `.FWHM`)), and in the multi-site routines (particularly handling of background).

In addition, due to requests from users, two additional features were added:

1. The ability to calculate the dipole transition matrix
2. The ability to add a residual (background) susceptibility (`chi0`) to the susceptibility calculations.

The first feature is purely Python, the second involved some C++ changes.

Most of the bugs were in the Python interface (e.g. the arbitrary `J` calculations was mistakenly removed in the Python interface but the C++ remained), but there were some C++ changes (particularly handling the peak widths) needed.

**To test:**

Code review and check automated tests run ok.

Test scripts and related datafiles are in the following zip:
[cf_test_scripts.zip](https://github.com/mantidproject/mantid/files/1744795/cf_test_scripts.zip)
Unzip and add the `data` folder to the Mantid data path and `reduction_scripts` to the Mantid scripts path and run. 

A new system test based on the `cf_doc.py` file has been added to check that the python syntax works (it runs some fits but the results are not checked - the unit tests are used for checked the numerical validity of the computations).

Fixes #20990

**Release Notes** 


Release notes included.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
